internal/eval/fold.go

```diff
@@ -0,0 +1,253 @@
+package eval
+
+import (
+	"fmt"
+	"slices"
+
+	"github.com/cedar-policy/cedar-go/internal/ast"
+	"github.com/cedar-policy/cedar-go/types"
+)
+
+// foldPolicy takes in a given policy and attempts as much constant folding as possible.
+// It is not given any environment (entities, parc) so any references of those sort will
+// stop the folding.  The kinds of things that this will fold:
+//
+//   - 1+1 -> 2
+//   - [1,2,3].contains(2) -> true
+//   - a nodes set [1,2,3] -> a value node of the set [1,2,3]
+//   - an extension call to Decimal("42") -> a value node of the Decimal("42")
+//
+// Expressions that will cause errors are not folded, for example, `Decimal("hello")` will
+// remain as an extension call and not be folded into a Decimal value.
+func foldPolicy(p *ast.Policy) *ast.Policy {
+	if len(p.Conditions) == 0 {
+		return p
+	}
+	p2 := *p
+	p2.Annotations = slices.Clone(p.Annotations)
+	if p.Conditions != nil { // preserve nility for test purposes
+		p2.Conditions = make([]ast.ConditionType, len(p.Conditions))
+		for i, c := range p.Conditions {
+			p2.Conditions[i] = ast.ConditionType{Condition: c.Condition, Body: fold(c.Body)}
+		}
+	}
+	return &p2
+}
+
+// NOTE: nodes is modified in place, so be sure to send unique copy in
+func tryFold(nodes []ast.IsNode,
+	mkEval func(values []types.Value) Evaler,
+	mkNode func(nodes []ast.IsNode) ast.IsNode,
+) ast.IsNode {
+	var values []types.Value
+	ok := true
+	for i, n := range nodes {
+		n = fold(n)
+		nodes[i] = n
+		if !ok {
+			continue
+		}
+		if v, vok := n.(ast.NodeValue); vok {
+			values = append(values, v.Value)
+			continue
+		}
+		ok = false
+	}
+	if ok {
+		eval := mkEval(values)
+		v, err := eval.Eval(nil)
```

Comment:
It's a little hard for me to understand how we know it's safe to pass in `nil` here. Could you document why it's okay?

Reply:
I _think_ it's because we know that `mkEval` is only going to ever return an `errorEval` (which doesn't look at the `Env`) or an `Evaler` whose arguments consist wholly of `types.Value`s and therefore don't need the `env`.